crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -177,6 +178,30 @@ impl Solution {
         Settlement::encode(self, auction, eth, simulator).await
     }
 
+    pub fn with_protocol_fees(self) -> Self {
+        let trades = self
+            .trades
+            .into_iter()
+            .map(|trade| match &trade {
+                Trade::Fulfillment(fulfillment) => match fulfillment.order().kind {
+                    order::Kind::Market | order::Kind::Limit { .. } => {
+                        let uniform_sell_price =
+                            self.prices[&fulfillment.order().sell.token.wrap(self.weth)];
+                        let uniform_buy_price =
+                            self.prices[&fulfillment.order().buy.token.wrap(self.weth)];
+                        fulfillment
+                            .add_protocol_fee(uniform_sell_price, uniform_buy_price)
+                            .map(Trade::Fulfillment)
+                            .unwrap_or(trade)
```

Comment:
I'm pretty sure we shouldn't fail silently here. We might not even want to continue with this solution as it could potentially cause significant damage if a solver doesn't withhold the correct amount of fee.

Reply:
Yes, was not sure about this, will add error handling.

Reply:
Added error handling